yandextank/plugins/DataUploader/client.py

```diff
@@ -118,49 +119,48 @@ def filter_headers(headers):
                 del (headers[h])
         return headers
 
-    def __send_single_request(self, req, trace=False):
-        p = self.session.prepare_request(req)
-        request_id = self.ids.next()
+    def __send_single_request(self, request, request_id, trace=False):
+        request.headers[self.REQUEST_ID_HEADER] = request_id
+        p = self.session.prepare_request(request)
         if trace:
-            logger.debug("""
+            logger.debug(self.format_request_info(p, request_id))
+        resp = self.session.send(p, timeout=self.connection_timeout)
+        if trace:
+            logger.debug(self.format_response_info(resp, request_id))
+        if resp.status_code in [500, 502, 503, 504]:
+            raise self.NotAvailable(
+                request=self.format_request_info(p, request_id),
+                response=self.format_response_info(resp, request_id))
+        elif resp.status_code == 410:
+            raise self.StoppedFromOnline
+        elif resp.status_code == 423:
+            raise self.UnderMaintenance
+        else:
+            resp.raise_for_status()
+            return resp
+
+    @staticmethod
+    def format_request_info(request, request_id):
+        return """
         Request:
             id: {}
             method: {}
             url: {}
             headers: {}
-            body: {}""".format(request_id, p.method, p.url, p.headers, p.body))
-        resp = self.session.send(p, timeout=self.connection_timeout)
-        if trace:
-            logger.debug("""
+            body: {}""".format(request_id, request.method, request.url, request.headers,
+                               request.body.replace('\n', '\\n'))
+
+    def format_response_info(self, resp, request_id):
+        return """
         Response:
             id: {}
             elapsed time: {}
             reason: {}
             status code: {}
             headers: {}
             content: {}""".format(request_id, resp.elapsed.total_seconds(),
-                                  resp.reason, resp.status_code, self.filter_headers(resp.headers), resp.content))
-        if resp.status_code in [500, 502, 503, 504]:
```

Comment:
add 'and not trace' condition to avoid duplicate data in log?

Reply:
@fomars